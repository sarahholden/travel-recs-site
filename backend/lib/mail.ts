import 'dotenv/config';
import {
  createTransport,
  getTestMessageUrl,
  SentMessageInfo,
} from 'nodemailer';

const transport = createTransport({
  host: process.env.MAIL_HOST,
  port: 587,
  auth: {
    user: process.env.MAIL_USER,
    pass: process.env.MAIL_PASS,
  },
});

function makeANiceEmail(text: string) {
  return `
    <div style="
      border: 1px solid black;
      padding: 20px;
      font-family: sans-serif;
      line-height: 2;
      font-size: 20px;
    ">
      <h2>Hello There!</h2>
      <p>${text}</p>
      <p>¬©Ô∏è, Sarah Holden</p>
    </div>
  `;
}

type MailResponse = {
  accepted?: string[] | null;
  rejected?: string[] | null;
  envelopeTime: number;
  messageTime: number;
  messageSize: number;
  response: string;
  envelope: {
    from: string;
    to?: string[] | null;
  };
  messageId: string;
};

export async function sendPasswordResetEmail(
  resetToken: string,
  to: string
): Promise<void> {
  const info = (await transport.sendMail({
    to,
    from: 'test@example.com',
    subject: 'Your Sick Fits password reset token!',
    html: makeANiceEmail(`Your Password reset token is here!
      <a href="${process.env.FRONTEND_URL}/reset?token=${resetToken}">Click here to reset</a>
    `),
  })) as MailResponse;

  if (process.env.MAIL_USER.includes('ethereal.email')) {
    /* eslint no-console: 0 */
    console.log(
      'üì¨ Message Sent! Preview it at',
      getTestMessageUrl(info as SentMessageInfo)
    );
  }
}
