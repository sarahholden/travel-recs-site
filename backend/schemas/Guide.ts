import { list } from '@keystone-next/keystone/schema';
import { text, relationship, select } from '@keystone-next/fields';

export const Guide = list({
  fields: {
    name: text({ isRequired: true }),
    location_info: text(),
    status: select({
      options: [
        { label: 'Draft', value: 'DRAFT' },
        { label: 'Available', value: 'AVAILABLE' },
      ],
      defaultValue: 'DRAFT',
      ui: {
        displayMode: 'segmented-control',
        createView: { fieldMode: 'hidden' },
      },
    }),
    destinations: relationship({ ref: 'Destination.guide', many: true }),
  },
});
