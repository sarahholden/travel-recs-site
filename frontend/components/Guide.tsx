import { Guide as guideType } from '../types/generated-queries';

type GuideProps = { guide: guideType };

export default function Guide({ guide }: GuideProps) {
  return (
    <div>
      {guide.name}

    </div>
  )
}