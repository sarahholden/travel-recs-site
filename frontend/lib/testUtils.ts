import casual from "casual";
// import { PAGINATION_QUERY } from '../components/Pagination';

// seed it so we get consistent results
casual.seed(777);

const fakeGuide = () => ({
  // __typename: 'Item',
  id: "abc123",
  user: null,
  altText: "New York City Skyline",
  destinations: [],
  image: {
    publicUrlTransformed: "dog.jpg",
  },
  location_info: "The biggest apple in existance",
  name: "New York City",
  status: "AVAILABLE",
});

const fakeUser = (overrides) => ({
  __typename: "User",
  id: "4234",
  name: casual.name,
  email: casual.email,
  permissions: ["ADMIN"],
  orders: [],
  cart: [],
  ...overrides,
});

const fakeOrderItem = () => ({
  __typename: "OrderItem",
  id: casual.uuid,
  image: {
    image: `${casual.word}.jpg`,
  },
  name: casual.words(),
  price: 4234,
  quantity: 1,
  description: casual.words(),
});

const fakeOrder = () => ({
  __typename: "Order",
  id: "ord123",
  charge: "ch_123",
  total: 40000,
  items: [fakeOrderItem(), fakeOrderItem()],
  createdAt: "2022-12-11T20:16:13.797Z",
  user: fakeUser(),
});

const fakeCartItem = (overrides) => ({
  __typename: "CartItem",
  id: "omg123",
  quantity: 3,
  product: fakeGuide(),
  user: fakeUser(),
  ...overrides,
});

// Fake LocalStorage
class LocalStorageMock {
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return this.store[key] || null;
  }

  setItem(key, value) {
    this.store[key] = value.toString();
  }

  removeItem(key) {
    delete this.store[key];
  }
}

function makePaginationMocksFor(length) {
  return [
    {
      request: { query: PAGINATION_QUERY },
      result: {
        data: {
          _allProductsMeta: {
            count: length,
          },
          itemsConnection: {
            __typename: "aggregate",
            aggregate: {
              count: length,
              __typename: "count",
            },
          },
        },
      },
    },
  ];
}

export {
  makePaginationMocksFor,
  LocalStorageMock,
  fakeGuide,
  fakeUser,
  fakeCartItem,
  fakeOrder,
  fakeOrderItem,
};
