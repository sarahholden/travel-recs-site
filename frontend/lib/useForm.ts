import { useState } from "react"

type GuideInputsType = {
  name?: string;
  location_info?: string;
}

type ElementPropsType = {
  name: string;
  type: string;
  value: string | number | File
}

export default function useForm(initial: GuideInputsType = {}) {
  const [inputs, setInputs] = useState(initial)

  
  function handleChange(e: React.ChangeEvent<HTMLInputElement> & React.ChangeEvent<HTMLTextAreaElement>) {
    let {name, type, value}: ElementPropsType = e.target;

    if (type === 'number') {
      value = parseInt(value)
    }

    if (type === 'file' && e?.target.files) {
      value = e.target.files[0]
    }

    setInputs({
      ...inputs,
      [name]: value
    })
  }

  function resetForm() {
    setInputs(initial);
  }

  function clearForm() {
    const blankState = Object.fromEntries(Object.entries(inputs).map(([key, value]) => [key, '']))
    setInputs(blankState)
  }

  return { inputs, handleChange, resetForm, clearForm }
  
}