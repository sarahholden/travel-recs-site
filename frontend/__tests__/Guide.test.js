import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/react-testing';
import Guide from '../components/Guide';
import { fakeGuide } from '../lib/testUtils';

const guide = fakeGuide();

describe('<Guide/>', () => {
  it('renders out the guide image', () => {
    const { container, debug } = render(
      <MockedProvider>
        <Guide guide={guide} />
      </MockedProvider>
    );
    const link = container.querySelector('a');
    expect(link).toHaveAttribute('href', '/guide/abc123');
    expect(screen.getByText('New York City')).toBeInTheDocument();
    // debug(link);
  });

  it('Renders and matches the snapshot', () => {
    const { container, debug } = render(
      <MockedProvider>
        <Guide guide={guide} />
      </MockedProvider>
    );
    expect(container).toMatchSnapshot();
  })
})