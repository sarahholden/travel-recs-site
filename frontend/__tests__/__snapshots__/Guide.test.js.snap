// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Guide/> Renders and matches the snapshot 1`] = `
<div>
  <article
    class="Guide__GuideStyles-sc-zh7nrp-0 dzznjp"
  >
    <div
      class="image-wrap"
    >
      <a
        href="/guide/abc123"
      >
        <img
          alt="New York City Skyline"
          src="dog.jpg"
        />
      </a>
    </div>
    <h2>
      <a
        href="/guide/abc123"
      >
        New York City
      </a>
    </h2>
    <button
      aria-label="Add To Favorites"
    >
      Heart
    </button>
    <p>
      The biggest apple in existance
    </p>
  </article>
</div>
`;
